api/src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java

```diff
@@ -0,0 +1,206 @@
+import java.io.OutputStream;
+import java.util.Map;
+import java.util.TreeMap;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import jenkins.model.ArtifactManager;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.tools.tar.TarEntry;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+
+/**
+ * Manages per-build stashes of files.
+ * Unlike artifacts managed by {@link ArtifactManager}, stashes:
+ * <ul>
+ * <li>Are expected to be transferred to other workspaces during the build.
+ * <li>Generally are discarded when the build finishes.
+ * <li>Are not exposed as part of the build outside Jenkins, for example via REST.
+ * <li>Are stored in an archive format with a simple name, not necessarily related to filenames.
+ * </ul>
+ */
+public class StashManager {
+
+    /**
+     * Saves a stash of some files from a build.
+     * @param build a build to use as storage
+     * @param name a simple name to assign to the stash (must follow {@link Jenkins#checkGoodName} constraints)
+     * @param workspace a directory to use as a base
+     * @param includes a set of Ant-style file includes, separated by commas; null/blank is allowed as a synonym for {@code **} (i.e., everything)
+     * @param excludes an optional set of Ant-style file excludes
+     */
+    @SuppressFBWarnings(value="RV_RETURN_VALUE_IGNORED_BAD_PRACTICE", justification="fine if mkdirs returns false")
+    public static void stash(@Nonnull Run<?,?> build, @Nonnull String name, @Nonnull FilePath workspace, @Nonnull TaskListener listener, @CheckForNull String includes, @CheckForNull String excludes) throws IOException, InterruptedException {
+        Jenkins.checkGoodName(name);
+        File storage = storage(build, name);
+        storage.getParentFile().mkdirs();
+        if (storage.isFile()) {
+            listener.getLogger().println("Warning: overwriting stash ‘" + name + "’");
+        }
+        OutputStream os = new FileOutputStream(storage);
+        try {
+            int count = workspace.archive(ArchiverFactory.TARGZ, os, new DirScanner.Glob(Util.fixEmpty(includes) == null ? "**" : includes, excludes));
+            if (count == 0) {
+                throw new AbortException("No files included in stash");
+            }
+            listener.getLogger().println("Stashed " + count + " file(s)");
+        } finally {
+            os.close();
+        }
+    }
+
+    /**
+     * Restores a stash of some files from a build.
+     * @param build a build used as storage
+     * @param name a name passed previously to {@link #stash}
+     * @param workspace a directory to copy into
+     */
+    public static void unstash(@Nonnull Run<?,?> build, @Nonnull String name, @Nonnull FilePath workspace, @Nonnull TaskListener listener) throws IOException, InterruptedException {
+        InputStream is = new FileInputStream(storage(build, name));
```

Comment:
ideally a check and informative warning rather than just throwing a FileNotFoundException.
(the user won;t really know that /opt/jenkins/jobs/bob/builds/stashes/wibb1e.tar.gz was nescessarily a typos in the stash archive name?